contracts/dev-utils/contracts/src/OrderTransferSimulationUtils.sol

```diff
@@ -0,0 +1,152 @@
+    /// @return The index of the first failed transfer (or 4 if all transfers are successful).
+    function getSimulatedOrderTransferResults(
+        LibOrder.Order memory order,
+        address takerAddress,
+        uint256 takerAssetFillAmount
+    )
+        public
+        returns (OrderTransferResults orderTransferResults)
+    {
+        // Create input arrays
+        bytes[] memory assetData = new bytes[](4);
+        address[] memory fromAddresses = new address[](4);
+        address[] memory toAddresses = new address[](4);
+        uint256[] memory amounts = new uint256[](4);
+
+        // Transfer `takerAsset` from taker to maker
+        assetData[0] = order.takerAssetData;
+        fromAddresses[0] = takerAddress;
+        toAddresses[0] = order.makerAddress;
+        amounts[0] = order.takerAssetAmount;
+
+        // Transfer `makerAsset` from maker to taker
+        assetData[1] = order.makerAssetData;
+        fromAddresses[1] = order.makerAddress;
+        toAddresses[1] = takerAddress;
+        amounts[1] = order.makerAssetAmount;
+
+        // Transfer `takerFeeAsset` from taker to feeRecipient
+        assetData[2] = order.takerFeeAssetData;
+        fromAddresses[2] = takerAddress;
+        toAddresses[2] = order.feeRecipientAddress;
+        amounts[2] = order.takerFee;
+
+        // Transfer `makerFeeAsset` from maker to feeRecipient
+        assetData[3] = order.makerFeeAssetData;
+        fromAddresses[3] = order.makerAddress;
+        toAddresses[3] = order.feeRecipientAddress;
+        amounts[3] = order.makerFee;
+
+        // Encode data for `simulateDispatchTransferFromCalls(assetData, fromAddresses, toAddresses, amounts)`
+        bytes memory simulateDispatchTransferFromCallsData = abi.encodeWithSelector(
+            _SIMULATE_DISPATCH_TRANSFER_FROM_CALLS_SELECTOR,
+            assetData,
+            fromAddresses,
+            toAddresses,
+            amounts
+        );
+
+        // Perform call and catch revert
+        (, bytes memory returnData) = address(_EXCHANGE).call(simulateDispatchTransferFromCallsData);
+
+        bytes4 selector = returnData.readBytes4(0);
+        if (selector == ASSET_PROXY_DISPATCH_ERROR_SELECTOR) {
+            // Decode AssetProxyDispatchError and return index of failed transfer
+            (, bytes32 failedTransferIndex,) = decodeAssetProxyDispatchError(returnData);
+            return OrderTransferResults(uint8(uint256(failedTransferIndex)));
+        } else if (selector == ASSET_PROXY_TRANSFER_ERROR_SELECTOR) {
+            // Decode AssetProxyTransferError and return index of failed transfer
+            (bytes32 failedTransferIndex, ,) = decodeAssetProxyTransferError(returnData);
+            return OrderTransferResults(uint8(uint256(failedTransferIndex)));
+        } else if (keccak256(returnData) == _TRANSFERS_SUCCESSFUL_RESULT_HASH) {
+            // All transfers were successful
+            return OrderTransferResults.TransfersSuccessful;
+        }
```

Comment:
Nit - it could be good to add a catch-all `else` here to show that we've considered all cases + for future-proofing.